TripalCultivate/TripalCultivate
# G1.27 Unable to save project-based content types

Issue #27 

## Description

We are currently not able to create any project-based content type through the UI because Tripal doesn't know that the project table has a type_id but our modules assume it does. This is because our modules are dependant on Chado v.1.3.3.013 whereas Tripal only supports chado 1.3 right now.

## Fix

The fix included in this PR is TEMPORARY until Tripal PR https://github.com/tripal/tripal/pull/1937 is resolved. It is not recommended but it will work.

Essentially, we are:
1. Install Chado v.13 using Tripal
2. Applying the SQL update commands to chado to update it to 1.3.3.013 outside of Tripal and without Tripal knowing.
3. Replacing Tripal's internal cache defining what Chado 1.3 is (i.e. `tripal_chado/chado_schema/chado_schema-1.3.yml`) with a cached version defining Chado 1.3.3.013 and letting Tripal think this new cached definition is for Chado 1.3.

This way our docker is using Chado 1.3.3.013 and despite Tripal thinking that we are using 1.3, our database matches what Tripal thinks it should.

## Testing

1. Create a new docker and container on this branch and ensure it does so without error.
2. Confirm that the chado installed has a type_id in the project table (i.e. that it is Chado v1.3.3.013):
```
drush sql:query "\d+ testchado.project"
```
> project_id      bigint          not null        nextval('testchado.project_project_id_seq'::regclass)   plain
> name    character varying(255)          not null                extended
> description     text                            extended
> type_id bigint                          plain                   An optional cvterm_id that specifies what type of project this record is.  Prior to 1.4, project type was set with an projectprop.
3. Confirm that Tripal thinks Chado 1.3 matches what we know is 1.3.3.013. This should not return any differences.
```
diff /var/www/drupal/web/modules/contrib/tripal/tripal_chado/chado_schema/chado_schema-1.3.yml /var/www/drupal/web/modules/contrib/TripalCultivate/config/sql/chado_schema-1.3.3.013.yml
```
4. Confirm that we can create project-based content types and they save their type in the project table just like they should.
   - Go to Admin > Tripal > Content > Add Tripal Content and select any "Research Experiment" type
   - Fill out all the required fields --basically just add a name-- and click save.
   - It should save without error
   - Finally select from the project table and confirm it has a type_id that matches the type of experiment you entered.
   ```
   drush sql:query "SELECT p.type_id, cvt.name as type, p.name FROM testchado.project p LEFT JOIN testchado.cvterm cvt ON cvt.cvterm_id=p.type_id"
   ```
   > 189    field study     Characterize soil microbiome associated with Lentil-Wheat in rotation across 2024-2027
